### DISCUSSION TITLE
Multi-interpreter refactoring roadmap

### DISCUSSION THREAD

**Discussion by ntoll** [pyscript/pyscript#1098] [Proposals]
@ntoll and @madhur-tandon spent a couple of hours reviewing both the PyScript and [MicroPyScript](https://github.com/pyscript/micropyscript) abstractions of multiple interpreters. The following discussion points are the result of this investigation:

- [madhur] Use Typescript for MicroPyScript
- [nicholas/madhur] PyScript currently reveals lots of Pyodide implementation details. This has to go away. E.g.
    - Loading of Python modules/packages: we reveal the several ways in which Pyodide does this in our base class. We should abstract this behaviour away into a single method, that the child class for Pyodide handles in the appropriate way, as would the child class for any other interpreter depending on how the specific interpreter manages this aspect of things.
    - Turn `loadPackage`/`installPackage` into just a single function.
    - `invalidate_module_path_cache` is a sticking plaster for Pyodide. We shouldn’t have that in the base class.
    - `this.globals` simply doesn’t exist except in Pyodide. We shouldn’t make this part of the base class.
    - `registerJSModule` is also a Pyodide only feature. Again, this shouldn’t be a part of the base class.
- [madhur/nicholas] Event driven mechanism for lifecycle? Using events to coordinate the unfolding of PyScript’s startup will ensure certain things can only happen in sequence. E.g. `onA(event)` will only be called when the “`A`” event is dispatched.
- [nicholas] It’s not clear to me how `stdin`/`stdout` connects with the underlying interpreter.
- [nicholas] Do we need the `name` and `lang` attributes..? They don’t add anything as far as I can tell.

**Comment by ntoll** [comment]
Here's what I hope:

* Constructive technical discussion of the above 👆 .
* Leading to clearly defined tickets any of us could pick up.
* We make the changes.
* It just works ™️ .

😛

**Comment by FabioRosado** [comment]
Thank you for opening this discussion! Funny enough I was thinking about the multi-interpreter this morning when taking the kiddo to the nursery 😄 

I think you and Madhur touched on a lot of the things I was concerned about, the biggest one being the runtime abstract class containing pyodide specific methods.

+1 to abstracting as much as we can! 

In the past, I have used the [cloudbridge](https://github.com/CloudVE/cloudbridge) to abstract interaction with multi-cloud providers. The code base was building and abstraction on top of cloudbridge's abstraction which made things confusing to understand and follow. It also meant that we had to build empty methods for 2 out of 3 cloud providers because only AWS supported a specific thing.

Just wanted to share this case, I'm sure we will be able to create a good abstraction without us having to create 'filler' methods because Pyodide supports x but Micropython (or others) doesn't 😄 

Regarding `this.globals` I think mostly we use this to reference python variables/funcs in javascriptland. Wouldn't this be useful depending on whichever interpreter we use?

**Reply by ntoll** [reply]
Sadly the `globals` is a Pyodide only feature.

**Reply by FabioRosado** [reply]
I see, would it be hard to abstract for micropython? 

Apologies for the silly question, if you want to reference a python function or variable in micropython does it just work? 😄

**Reply by ntoll** [reply]
There are no silly questions. :+1: And what you ask is important.

If you want to call a function in MicroPython you just do:

```JavaScript
mp_js_do_str("my_python_function('hello')");
```

Basically, you just ask MicroPython to evaluate your function call. But since there is no Python/JavaScript broker layer in MicroPython (nor in CPython for WASM either), and this is a Pyodide only "feature", then you can't access Python from JavaScript (and vice versa).

I'm currently writing up the PolyPlug docs, which also addresses this in an interpreter agnostic way (i.e. it'll work for all versions of Python because it's based on message passing rather than requiring a bespoke implementation-specific JS broker that depends on emscripten).

**Reply by ntoll** [reply]
It's worth pointing out that there are (as always) trade-offs with each approach and which trade-off you accept depends upon what you're trying to achieve. Let many flowers bloom. :tulip:

**Reply by JeffersGlass** [reply]
For what it's worth, [pyodide.globals](https://pyodide.org/en/stable/usage/api/js-api.html?highlight=globals#pyodide.globals) is a proxy for the Python global namespace [i.e. the return value from `globals()`]. If `globals()` works in Micropython, I wonder if we might implement be able to implement an `interpeter.globals` property that would delegate object access to the relevant Interpreter in a clean way?

**Reply by ntoll** [reply]
@JeffersGlass that's a good idea, and MicroPython does have a `globals`. If I do:

```python
print(globals())
```
I get this to `stdout`:
```python
{'__name__': '__main__'}
```

But I foresee problems, there's no JS/Python broker layer for MicroPython, so (for example), if there's a `foo` in `globals` that is an instance of some custom Python class, it's hard to know how to represent the `foo` object in the JS side of things without re-implementing a complete broker layer. I'd also argue that this is actually a feature (not a bug) since translating seamlessly between Python <-> JS is non-trivial and I actually prefer to keep these two "realms" separate except for a clearly defined message protocol that exposes aspects of each world to the other without "Pythonic" bleeding into JavaScript and vice versa. Trade-offs again... :wink: 

Of course, this _could_ be done via a JSON message in PolyPlug... e.g. we define a message type that defines a Python object, it lists the object's attributes and includes simple attribute values. But to my mind (at this moment in time) this feels like an abstraction too far, and perhaps something a framework higher up the stack could implement at a future date (i.e. PolyPlug should only concern itself with message passing -- I'm anxious to keep things small, simple, well defined and avoid scope creep, IYSWIM).

Also, async-y things from each world in the other... oh my. In Python how ought we to handle a result that is a JS promise..? Or what about blocking network calls in Python where JS is expecting something async to happen..? Since I'm a Winnie the Pooh sort of a developer (very small brain), this makes my head hurt just thinking about the potential mess. :grinning: To quote Homer Simpson... "if something's too hard to do, it's probably not worth doing anyway." :rofl: (I'm semi-serious here.)

**Comment by antocuni** [comment]
I think that yesterday someone asked my opinion on this, so let me chime in here.
First of all, I'd like to clearly define what we mean by "multi-interpreter":
1. run a pyscript page using a different interpreter than pyodide
2. run a pyscript page which can use different interpreters *at the same time*

I think that (2) is a very complex problem so I assume that we are talking about (1), please correct me otherwise.

### What and how to abstract

Honestly, the discussion above sounds a bit too vague to me, to the point that I don't really understand what is the proposed solution. It's not enough to say "we should remove feature XXX", we need to understand what would be the substitute of XXX.

Moreover, let's not forget about the [Law of leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction). The TL;DR version of it is that no matter how hard we try, it's impossible to write the "perfect" abstraction.
This also matches my experience: if one tries to build an abstraction upfront, without knowing which will be the concrete implementations, they will inevitably fail.

All of this to say that IMHO the correct way of tackling this issue is to look at the concrete problem, which in this case it's not "multi-interpreter" but "pyodide and micropython".

So, IMHO the correct questions to ask are:
1. what is the minimal set of features that pyscript needs from the interpreter to execute "useful code"?
2. how can micropython implement these features?
3. I have the feeling that at least for the time being, micropython will offer less features than pyodide. How do we want to handle graceful degradation? 

To explain what I mean in the point 3, consider these two examples:
1. pyscript supports `<script type="importmap">`, which is implemented by calling `registerJsModule`.
https://github.com/pyscript/pyscript/blob/e1b4415193e3c18f6fc5ccbccff712281dea7f48/pyscriptjs/tests/integration/test_importmap.py#L8-L41
https://github.com/pyscript/pyscript/blob/e1b4415193e3c18f6fc5ccbccff712281dea7f48/pyscriptjs/build/plugins/importmap.ts#L49

MicroPython doesn't offer this feature: what do we do if an user tries to use an `importmap`? We can ignore it, or we can raise an error. (I don't want an answer to this precise question now, but this is the sort of questions that we need to take into account)

2. python plugins: the current implementation of python plugins makes heavy use of the pyodide FFI. We can try to refactor it completely to avoid the FFI, or declare that python plugins are not supported by micropython (for now).

Once we have the answers to these question, we can invent an abstract interface which can be implemented reasonably for both interpreters.

But before diving into that, it's worth talking more about the FFI.

### FFI vs message passing

This is the elephant in the room, IMHO. I already expressed my concerns to some of you privately, but it's probably a good idea to write them down.

Pyodide offers the FFI interface, aka Python<->JS bridge. It makes it possible to pass JS objects into Python and Python objects into JS. The current codebase of PyScript relies on it heavily, but micropython doesn't offer it.
We have two options:
1. refactor the pyscript codebase to avoid relying on the FFI
2. add some sort of FFI to micropython

Solution (1) is basically the one adopted by Nicholas in MicroPyScript. If I understand correctly, he uses message passing to simulate "calls" from python to JS and viceversa. This approach has a *huge* advantage: it automatically works also for webworkers.
However, it has also big disadvantages. The biggest limitation of message passing is that if you do it naively you can pass around only primitive types (i.e., numbers and strings): you cannot arbitrary objects across the language border, everything is immutable, and callbacks are impossible (i.e., passing a python function to JS in a way that JS can call it later). It is probably possible to emulate all these things in a message-passing environment, but it's not trivial and requires a lot of thinking.

Solution (2) doesn't exist yet, but I'm pretty sure that we will need a micropython FFI eventually: implementing it means to modify the micropython code base to add a new C extension, reusing the same ideas (and possibly some of the code) which are used to implement the equivalent thing in pyodide. My gut feeling is that it is probably simpler than implementing the full message-passing logic which I described above, but I don't know the micropython codebase so I might be wrong here.
Solution (2) has several advantages over (1):
1. it automatically gives you access to the DOM, no need for polyplug
2. it automatically gives you access to non-DOM JS libraries (e.g. d3.js), which we will eventually need anyway
3. it makes micropython closer to pyodide in terms of features: this means that the pyscript codebase needs to "abstract less", which probably results in a better and simpler codebase
4. it is work that we need to do *anyway* eventually. Better to do it earlier and take advantage of it, instead of spending a lot of time into temporary workarounds.

### Summary

Let me try to write a summary of what I'm trying to convey here:

1. pyscript needs to accommodate the differences between pyodide and micropython: we can either introduce more complexity into the pyscript codebase, or try to reduce the differences between the two

2. working around the absence of a micropython FFI is as hard (if not harder) as implementing it "properly". IMHO implementing it properly is a better usage of our time and resources because it unlocks much more features and reduces the complexity of pyscript itself.

3. message passing is not evil. We will need it for webworkers, but this is a slightly different problem which probably needs a slightly different solution. If we want to try message passing (and I think we should), it's probably a better idea to start with pyodide/webworkers than with micropython.

**Comment by fpliger** [comment]
Now, this sounds like a cool party I don't want to miss. 😝  

As many have expressed, this is not going to be an easy task to scope without getting our hands dirty. I'll intentionally try to get straight to the point on how I think we should approach this and get less into the technical details of some of the options.

In general, this is the type of problem/task that calls for a strong divide-and-conquer strategy. It's way simpler to start cutting some pieces out, changing things, etc.. without actually also implementing support for MicroPython. Actually, a lot of the things we need to do were already improvements that we wanted to do independently of MicroPython, so it just becomes a nice forcing function in that aspect...

> PyScript currently reveals lots of Pyodide implementation details. This has to go away.

Yup. As I mentioned above, this is something very actionable even now that can happen at anytime (actually, the sooner the better)

> [madhur/nicholas] Event driven mechanism for lifecycle? Using events to coordinate the unfolding of PyScript’s startup will ensure certain things can only happen in sequence. E.g. onA(event) will only be called when the “A” event is dispatched.

This is also nice and, imho, we should explore that route but(!) this also is independent from "pure" MicroPython support and can happen at any time. 

> [nicholas] It’s not clear to me how stdin/stdout connects with the underlying interpreter.

Can you clarify a bit more around the specifics here?

> [nicholas] Do we need the name and lang attributes..? They don’t add anything as far as I can tell.

We don't "need" it... but, this also, is really just cosmetic compared to the big effort to support MicroPython

> I think that yesterday someone asked my opinion on this, so let me chime in here.
> First of all, I'd like to clearly define what we mean by "multi-interpreter":
> 
> run a pyscript page using a different interpreter than pyodide
> run a pyscript page which can use different interpreters at the same time
> I think that (2) is a very complex problem so I assume that we are talking about (1), please correct me otherwise.

As expressed a few times, ideally, I'd love to support 1 and 2 but, as @antocuni , (2) is definitely more complex and, imho, needs to happen after we have learned lessons from (1). So, I would put it on hold and only consider 1 for now.

> Honestly, the discussion above sounds a bit too vague to me, to the point that I don't really understand what is the proposed solution. It's not enough to say "we should remove feature XXX", we need to understand what would be the substitute of XXX.

+1 

> Moreover, let's not forget about the [Law of leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction). The TL;DR version of it is that no matter how hard we try, it's impossible to write the "perfect" abstraction.
> This also matches my experience: if one tries to build an abstraction upfront, without knowing which will be the concrete implementations, they will inevitably fail.
> 
> All of this to say that IMHO the correct way of tackling this issue is to look at the concrete problem, which in this case it's not "multi-interpreter" but "pyodide and micropython".

+1

> So, IMHO the correct questions to ask are:
> 
> 1. what is the minimal set of features that pyscript needs from the interpreter to execute "useful code"?
> 2. how can micropython implement these features?
> 3. I have the feeling that at least for the time being, micropython will offer less features than pyodide. How do we want to handle graceful degradation?

+1. Once we have 1 and 2 we can think about a good plan for graceful degradation. More on this below...


> To explain what I mean in the point 3, consider these two examples:
> 
> pyscript supports <script type="importmap">, which is implemented by calling registerJsModule.
> [pyscript/pyscriptjs/tests/integration/test_importmap.py]
> 
> https://github.com/pyscript/pyscript/blob/e1b4415193e3c18f6fc5ccbccff712281dea7f48/pyscriptjs/build/plugins/importmap.ts#L49
> MicroPython doesn't offer this feature: what do we do if an user tries to use an importmap? We can ignore it, or we can raise an error. (I don't want an answer to this precise question now, but this is the sort of questions that we need to take into account)
> 
> python plugins: the current implementation of python plugins makes heavy use of the pyodide FFI. We can try to refactor it completely to avoid the FFI, or declare that python plugins are not supported by micropython (for now).
> Once we have the answers to these question, we can invent an abstract interface which can be implemented reasonably for both interpreters.

... I think @antocuni  points above kind of give an answer alredy (considering my point earlier about divide and conquer) . The simplest option is to just declare that some features are not available in MicroPython for the time being. Once we have a list of these and can also evaluate the effort to actually add them to MicroPython we can make predictions around whether we'll add them in the future or not. Basically, what I'm trying to say is that, for MicroPython, we start small and build on top of it...

> Pyodide offers the FFI interface, aka Python<->JS bridge. It makes it possible to pass JS objects into Python and Python objects into JS. The current codebase of PyScript relies on it heavily, but micropython doesn't offer it.
> We have two options:
> 
> refactor the pyscript codebase to avoid relying on the FFI
> add some sort of FFI to micropython

These 2 options are not mutually exclusive. It's totally realistic to support 1 (which is mostly about supporting message passing between workers and main process) and 2...  We can add support for them at different times and, once we have both options available, allow users to pick their option.

> The biggest limitation of message passing is that if you do it naively you can pass around only primitive types (i.e., numbers and strings): you cannot arbitrary objects across the language border, everything is immutable, and callbacks are impossible (i.e., passing a python function to JS in a way that JS can call it later). It is probably possible to emulate all these things in a message-passing environment, but it's not trivial and requires a lot of thinking.

We don't *need* to support that. Supporting message passing with only primitive types support can take us a long way...

> Let me try to write a summary of what I'm trying to convey here:
> 
> pyscript needs to accommodate the differences between pyodide and micropython: we can either introduce more complexity into the pyscript codebase, or try to reduce the differences between the two
> 

Not sure I understand this statement. @antocuni , can you clarify?

> working around the absence of a micropython FFI is as hard (if not harder) as implementing it "properly". IMHO implementing it properly is a better usage of our time and resources because it unlocks much more features and reduces the complexity of pyscript itself.

What I read here is that implementing a proper FFI for MicroPython is probably as hard as implementing message passing. Is this correct? If so, I'm not entirely convinced about it, given the amount of working code that @ntoll already has...


> message passing is not evil. We will need it for webworkers, but this is a slightly different problem which probably needs a slightly different solution. If we want to try message passing (and I think we should), it's probably a better idea to start with pyodide/webworkers than with micropython.

Yes! Tbh, that's what I believe as well (they are different features that, in some cases, can provide solution to same problems) but, untile now, you seemed to have a different opinion.

Either way, I think you are right about starting support for it with pyodide/webworkers rather than MicroPython first (that goes back to my point about dividing and conquering the bigger problem here...)

**Comment by ntoll** [comment]
Hi Folks,

:wave: Thank you @antocuni and @fpliger for such stimulating and thoughtful comments.

My TL;DR:

* These discussions are evidence of good engagement in a messy problem space, and by taking part in this thoughtful reflection, we're going to save ourselves a bunch of problems in the future.
* In general I agree with much of what you have both written. I'd only add that given the complexity of the situation (multi-interpreters, main-thread/web-worker, message-passing/FFI), a multi-faceted and "blended" approach is key. By that I mean we have several dimensions to the project that are now multi-faceted when once there was only a single (Pyodide based) solution. I'm definitely in the, "let many flowers bloom" camp, and I'm certainly NOT trying to promote one way over another. Rather, I see the current work we're putting into this discussion and related experiments as a way to clarify borders, overlaps and different approaches. I hope you agree, each has its place, we should cultivate each facet to its full potential, while also being honest about each approach's limitations. (Hence discussions like this one.)
* This won't be easy, will probably be quite frustrating and I expect we'll get things wrong. But I'm also confident in the strength of this team, the thoughtfulness of the folks involved in the project and our self-awareness to reflect, refine and change as needed. We've got this. Let's get on with it.
* I have a proposal for how we define interpreters in our discussions, according to levels of support. See the very bottom.

To flesh the TL;DR out... here's some inline commentary to Antonio's and Fabio's responses:

## Antonio

### Scope

Couldn't agree more with you @antocuni with regard to clarification:

> 1. run a pyscript page using a different interpreter than pyodide
> 2. run a pyscript page which can use different interpreters at the same time
>
> I think that (2) is a very complex problem so I assume that we are talking about (1), please correct me otherwise.

Yes. That's the assumption I'm making. Although if we do (1) well, then (2) (should we go in that direction) will be easier to do. Note, I'm not saying (2) is easy, just that (1) could/might make it easier. My intuition is that "different interpreters" is going to be difficult to the extent that we should ask ourselves if the cost of tackling such complexity is worth the benefits (what are the benefits of running MicroPython AND Pyodide on the same page?).

I'd also like to add a (1.5) to your list: _multiple instances of the same interpreter running in different web-workers_. I'd add that we should _clearly ring-fence and constrain the scope of each interpreter's interaction with the main thread / DOM so they don't interfere with each other_.

### What to abstract

Again, couldn't agree more about leaky abstractions. I'd only add that, right now, we're very leaky in a Pyodide way. 

I think your approach of focusing on a concrete problem (abstract for both MicroPython AND Pyodide) is the way to go, and is essentially the approach I've been taking over in MicroPyScript - I've just had the advantage of starting from new.

> I have the feeling that at least for the time being, micropython will offer less features than pyodide. How do we want to handle graceful degradation?

This is a well made point and important question. 

I hope you don't mind, but I'd like to refine the point to, "MicroPython will offer **different** features to Pyodide" (for example, a much faster start-up time), so the question becomes, "how do we want to gracefully signal such differences?".

I think our notion of a "platform" is important here. If PyScript makes it easy to use Python in the browser, you have to ask yourself (as the user), "use Python for what?". Perhaps our line should be, PyScript, like a car manufacturer, allows you to choose different models of interpreter (vehicle) and customise depending on your use case and expected "road conditions". You wouldn't use a Fiat 500 for off-roading, and a 4x4 would be over-kill for dropping the kids to school. That sort of thing... and clearly one aspect of these discussions is to make sure we get our story straight.

You also mention `registerJsModule` and correctly point out that MicroPython doesn't support this feature. I'd say that's a feature, not a bug because not everyone will want nor require such close JS/Python integration. Which brings us to the [Heffalump](https://en.wikipedia.org/wiki/Heffalump). :elephant: 

### FFI vs Message Passing

Honestly, I hope we make sure our PyScript soil is fertile and well tilled enough that we let many flowers bloom. :+1:

You succinctly describe the FFI as:

> Pyodide offers the FFI interface, aka Python<->JS bridge. It makes it possible to pass JS objects into Python and Python objects into JS. 

Some folks will want this, many won't. But we should definitely make it available because it's so useful for folks using Pyodide.

I think my problem is when you explain the current situation as:

> The current codebase of PyScript relies on it heavily, but micropython doesn't offer it.

To continue my car analogy, that's like saying "My current Land Rover supports all wheel drive, but the Fiat 500 doesn't offer it".

I agree with your list of possible ways forward. I think this is a logical AND not an OR.

1. We should **NOT** rely on FFI, **AND**
2. We should add some sort of FFI to MicroPython.

Why? As more interpreters become available, option 1 means it ought to be relatively easy to support them with limited (message passing?) based support. If/when demand drives it, we should have a well trodden path, or perhaps even a framework already built, for bolting on some sort of FFI. It's a win/win (of sorts). :+1:

I'm not going to go into the :+1: / :-1: aspects of FFI vs message passing. To be honest, I think this sort of discussion simply shows they're different and both make trade-offs that mean the other has better features in different contexts. But that's the point... we want our PyScript platform to support these different contexts: clearly you shouldn't go off-roading in a Fiat 500, and if you need a nippy little car to get you through the city traffic, don't go for a 4x4.

But what we ought to do, and I think this :+1: / :-1: aspect of the conversation is helping us to get to, is *clearly define how different aspects line up to contexts and use cases*. Don't use MicroPython for heavy duty data science that requires a JS native presentation layer, and likewise, best of luck getting Pyodide to start in less than 20 seconds on a mobile device (both legitimate use cases we've discussed for PyScript).

I hope this makes sense. I'm not disagreeing with you, rather, I'm trying to tell the story in a slightly different way. :+1:

### Summary

Love this.

1. Reducing the differences between the two isn't trivial and might be a futile exercise in trying to get a Fiat 500 to behave like a 4x4. 
2. Unlocks more features? Like not working over web-workers? For this reason alone, I'd argue we also need message passing. Like I said, I see this aspect of our project as a logical AND, not an OR. Both FFI and message passing have their place (and PyScript is better because we have both).
4. Starting with Pyodide/web-workers excludes MicroPython from the process. PolyPlug, from the start, is designed to work with both, and in the main thread and across web workers. It also works already. Let many flowers bloom. :rose:

## Fabio

Love the "I'm going for the jugular" approach here. :grin: You make an excellent point when you say:

> In general, this is the type of problem/task that calls for a strong divide-and-conquer strategy. It's way simpler to start cutting some pieces out, changing things, etc.. without actually also implementing support for MicroPython. Actually, a lot of the things we need to do were already improvements that we wanted to do independently of MicroPython, so it just becomes a nice forcing function in that aspect...

I read that (and please correct me if I'm wrong) as, much of this stuff is orthogonal to supporting MicroPython.

You also asked for specifics about my stdin/stdout connection clarification. I think what I'm saying is that it's not obvious, when looking at the current PyScript code, how stdin/stdout interact with the underlying interpreter. This is, I believe, because Pyodide has allowed us to rely on FFI to bridge the gap between the browser and the interpreter, rather than via the Unix-y model of piping via stdin/stdout. Incidentally, I believe the Unix-y model of piping via stdin/stdout is perhaps the best shot we have at removing our dependency on FFI. But see below, in my proposal, for the implications of this. I'm not saying we shouldn't use FFI either. It's another "let many flowers bloom" situation. :rose: 

If we don't need it, then it probably shouldn't be there (`name` and `lang` attributes). Less code is good, no code is best. Code or attributes that don't exist, can't have or introduce bugs.

Good to see that we're aligned on the (1) vs (2) discussion.

Also, with the rest of your commentary Fabio, I think we're aligned in that many-flowers sort of a way. I also think your divide and conquer approach is a good one.

I think, after reading both replies, we're actually more aligned than it at first appears. :smile: The difficulty in our journey ahead is to make it clear how we handle the differences in capability... which leads me to my proposal.

## My Proposal for Discussions

To help us better discuss these things I'd like to propose some levels of capability (kinda like the ones used for self-driving cars). We can adjust and refine the specific levels, but as a first draft / straw man I propose the following.

When **explaining PyScript's interpreters** we introduce the notion of levels. The lower the level the less capable that particular version of PyScript is, **BUT** the more simple, smaller and easier to implement it is likely to be, thus clearly and unambiguously revealing the inevitable trade-offs. Folks can look at their use case, figure out the level they need and end up with a list of candidate interpreters to use in PyScript.

Straw man levels are listed below (please suggest updates!). Each level adds to the capabilities of the lower levels:

1. The interpreter evaluates code in a `<py-script>` tag, in the main thread with output coming via `stdout` and limited user input handled via `stdin`. (This is basically MicroPyScript pre-PolyPlug).
2. The interpreter uses simple standard message passing to interact with the browser/DOM. As a result it is able to run in both the main thread and web-workers. This uses `stdin/stdout` and on-the-fly `eval`-uation of code. (This is where I'm almost at with PolyPlug and my work on web-workers.)
3. The interpreter uses a _standard_ FFI to interact with JavaScript and the DOM in a clearly specified way. (Not necessarily the Pyodide FFI, which feels to me very Pyodide specific, but perhaps a more generic / universal version inspired by / forked from the great work of Pyodide.)
4. The FFI and message passing capabilities are integrated to allow appropriate FFI work to happen on web-workers.
5. The interpreter runs on a Delorean so we can travel back to the future to get the finished version of PyScript that will work for all people on all devices to tackle all computing problems.

I'm not sure about number (5), but I think numbers 1->4 are sensible in a "first steps" sort of a way.

Joking aside, as we go up the levels things become more complicated and "bigger". It's important to point out that "the 99%" (our shortcut phrase for, "folks who are non-technical") could probably get away with PyScript at level 2. If you're technical enough to know you need the FFI, you're probably not part of the 99%, if you see what I mean.

Hope this makes sense. I've really enjoyed responding to your excellent and thoughtful comments... really stimulating and thought provoking, which is exactly what good engineering should feel like IMHO. :rocket:

I'm really looking forward to where this goes... but in the first instance I think [my comment about my hopes](https://github.com/pyscript/pyscript/discussions/1098#discussioncomment-4600381), still stands, and we're currently at the first bullet point (_Constructive technical discussion_).

Onwards..!

**Comment by fpliger** [comment]
@ntoll thank you for the thorough and well-written response. I think we are very very aligned.  There's one main aspect where my POV differs:

> Joking aside, as we go up the levels things become more complicated and "bigger". It's important to point out that "the 99%" (our shortcut phrase for, "folks who are non-technical") could probably get away with PyScript at level 2. If you're technical enough to know you need the FFI, you're probably not part of the 99%, if you see what I mean.

I'm not convinced about this. I think that FFI access provides access to a lot of features [in the JS world] that are quite interesting to "the 99%", such as browser APIs to access lower-level system resources (camera, geolocation, etc..) and things like JS game engines, and other libraries that allow users to create "cool stuff". 

We may be able to access some (maybe all) of the above with level 2 but not really sure we can maintain the same API simplicity and make the choice between 2 and 3 almost transparent to the users.

What I will say is that, imho, we should really strive to keep 2 and 3 as simple and as similar as possible.


Quick answer to your question about:

> If we don't need it, then it probably shouldn't be there (name and lang attributes). Less code is good, no code is best. Code or attributes that don't exist, can't have or introduce bugs.

That idea of `name` and `lang` being there was mainly because I wanted to surface more information about an interpreter than just a link... it's probably a premature choice (quite early on in dev) when I had less context and not needed tbh.

**Reply by ntoll** [reply]
@fpliger ahhh... I must apologise for not making myself clearer... Lemme try again. :grin: 

What I had at the back of my mind was that PyScript at level 2 would give enough features that "the 99%" could find it useful. Clearly, they'll find FFI based things useful too... but I was trying to point out that perhaps at "level 2" we should also be focusing on the core intents, tasks and features that a "99%" type coder would find useful. For instance, a simple DOM manipulation API with perhaps some abstracted access to common web APIs provided by the browser (such as geolocation, canvas compass). I'm basically describing what PolyPlug can/will do. Furthermore, if you're the sort of developer who wants a D3 visualisation and access to other "funky" stuff, you definitely need FFI, but at the same time, you're probably not a "99%" coder either because you know about D3. :wink: 

Does that make sense..?

I was trying (unsuccessfully) to try to say something about the **focus** in terms of expected users for the different levels. The rule of thumb being that lower levels should concentrate on "the 99%", the higher levels, as PyScript becomes more powerful as it integrates closer to the underlying browser APIs and infrastructure, should enable and empower "power user" devs.

There's also another educational aspect that I had at the back of my mind, which I didn't mention, so I'll mention it now... learner context. 

It's far easier to explain to a beginner how messages work, than it is to explain the intricacies of an FFI (because you need to understand both JavaScript and Python and types and other engineer-y things). Message passing is simply a glorified "telephone call" analogy. I phone the browser from Python to tell it to display this stuff, and the browser calls me back when such-and-such a button is pressed. Again, my thinking is that the mental models needed for the lower levels should be more beginner friendly than the ones needed at the (more powerful and intricate) higher levels.

Ack on `name` and `lang`. We're refining and adjusting. That's just a part of coding.. :-)

**Reply by fpliger** [reply]
> @fpliger ahhh... I must apologise for not making myself clearer... Lemme try again. 😁

Absolutely no need to apologize! Communicating is hard :P

> What I had at the back of my mind was that PyScript at level 2 would give enough features that "the 99%" could find it useful. Clearly, they'll find FFI based things useful too... but I was trying to point out that perhaps at "level 2" we should also be focusing on the core intents, tasks and features that a "99%" type coder would find useful. For instance, a simple DOM manipulation API with perhaps some abstracted access to common web APIs provided by the browser (such as geolocation, canvas compass). I'm basically describing what PolyPlug can/will do. Furthermore, if you're the sort of developer who wants a D3 visualisation and access to other "funky" stuff, you definitely need FFI, but at the same time, you're probably not a "99%" coder either because you know about D3. 😉
> 
> Does that make sense..?

Yes but I guess I have mixed feelings about the above. Honestly, I think there is a place for both FFI and message passing and, if you are successful at what we are doing, the difference to the final users is not substantial in most cases but rather in some of the details. A huge part of the path for us to be successful is actually reducing complexity everywhere possible and that applies for both 2 and 3 above (and both have their role in supporting our user needs).

That same principle applies to another example you made above:

> if you're the sort of developer who wants a D3 visualisation and access to other "funky" stuff, you definitely need FFI, but at the same time, you're probably not a "99%" coder either because you know about D3. 😉

I really hope you are wrong about that my friend. My hope is that us/the community can build a higher level of these libraries so that they can actually become accessible to the 99%. I agree with you that the 99% demographics will not come looking for D3 viz but rather "cool viz" and through examples and plugins they could be able to find an example app that does what they are looking for and can copy and change values to fit their needs... Or something like that. One can dream... (and work hard to make it a reality) :)

> I was trying (unsuccessfully) to try to say something about the focus in terms of expected users for the different levels. The rule of thumb being that lower levels should concentrate on "the 99%", the higher levels, as PyScript becomes more powerful as it integrates closer to the underlying browser APIs and infrastructure, should enable and empower "power user" devs.

Yeah, +1 on focus in terms of expected users for the different levels. For some reason here (and I think it's mainly on the working since I'm pretty sure we are aligned on that ) we are seeing it flipped lol. I expect higher level APIs (more expressive and easy but maybe less powerful) to focus on the 99% and the lower level (harder but more powerful) to be more used by "power users".

> It's far easier to explain to a beginner how messages work, than it is to explain the intricacies of an FFI (because you need to understand both JavaScript and Python and types and other engineer-y things). Message passing is simply a glorified "telephone call" analogy. I phone the browser from Python to tell it to display this stuff, and the browser calls me back when such-and-such a button is pressed. Again, my thinking is that the mental models needed for the lower levels should be more beginner friendly than the ones needed at the (more powerful and intricate) higher levels.

Agreed on the analogy being easier to explain. At a higher level though, and I'm repeating myself here, I hope that we don't even have to explain the difference because many users "won't notice" and don't even care to know :)

Overall, I feel like we are very aligned and mostly fighting with terminology and putting words on a GH post

**Reply by ntoll** [reply]
@fpliger pretty much 100% agree with you when you say,

> Overall, I feel like we are very aligned and mostly fighting with terminology and putting words on a GH post

I think when we are sat in the same room, able to chat at length and not play "GitHub comment ping-pong" we'll get a huge amount done... including the important work of nurturing terminology, focus and so on.

FWIW - unless explicitly stated, it's never a logical "or" with me, but an "and" - just because I write "X is the case", doesn't mean "Y can't be the case" either. And yes, I agree with you WRT the "won't notice" point, but it's part of our job "to notice" _so they don't have to_ if you see what I mean (hence this [very rewarding] sort of discussion). :rocket:

**Comment by marimeireles** [comment]
Nice discussion, folks!

Going through some points that people brought up:

>We don't need to support that. Supporting message passing with only primitive types support can take us a long way...

Exactly! +1 for this, we can still execute code. That was my first attempt to webworkers, even though it limits a lot of what we can do because of the lack of possibility to manipulate the DOM. I think it's a good first step to implementing webworkers in PyScript =)
Just basic types, no FFI nor Polyplug integration. And if I understand Antonio's and Fabio's points (not sure about Nick) we're all converging to that idea.

>it automatically gives you access to the DOM, no need for polyplug

I don't understand this statement @antocuni. Are you saying that pyodide's FFI would give us access to the DOM from within a webworker?

About the FFI vs message passing discussion... I'm not sure if I follow here, so I appreciate some enlightment!
I think the reason for my confusion might be because people are talking about two message passing mechanisms?
1 is the one for webworkers, the API we need to agree on to pass content from the main to the worker thread. 2 is the mechanism developed in polyplug that essentially modifies the DOM tree from the worker, which is also working as FFI but AFAIU Pyodide's FFI doesn't do the same thing, it doesn't allow us to interact with the DOM from the worker thread.
So, to clarify, the discussion is: should we have an implementation that translates js<>py objects that relies on FFI Pyodide's design or in the polyplug json deserialization and serialization design? If this is the discussion we're having, I don't really see how it makes sense to have both.
But I might be just totally confused 😅

Other than that, pretty cool! Excited to see this discussion shaping up and I'm either agreeing with the last few comments or no strong opinions. ☀️

**Reply by ntoll** [reply]
> Just basic types, no FFI nor Polyplug integration. And if I understand Antonio's and Fabio's points (not sure about Nick) we're all converging to that idea.

Sure, that's "level 1" in my proposed system of levels, but in order to be able to manipulate the DOM and react to events you'd need at least "level 2" or more (with PolyPlug or equivalent providing this level). "Level 3" would be to add direct access to JS via an FFI. Does that make sense..?

As for two message passing things. I'm not sure I understand. A message is a message is a message, no matter if it's sent from a web-worker or dispatched via an event in the main thread (and I think that's the point, right?).

> So, to clarify, the discussion is: should we have an implementation that translates js<>py objects that relies on FFI Pyodide's design or in the polyplug json deserialization and serialization design? If this is the discussion we're having, I don't really see how it makes sense to have both.

I'm saying the opposite. I think we could and should have both, for all the reasons I explained in my response to Fabio and Antonio. See especially the discussion at the end about levels. This isn't a logical "or", but an "and" so PyScript can benefit from the benefits both approaches bring.

Does that make sense..?

**Reply by marimeireles** [reply]
Okay, as discussed in the meeting will wait for the opening of the issues! :)
